src/Workspaces/Core/Portable/ObsoleteSymbol/AbstractObsoleteSymbolService.cs

```diff
@@ -0,0 +1,176 @@
+            }
+        }
+
+        SyntaxToken ProcessNode(SemanticModel semanticModel, SyntaxNode node)
+        {
+            if (syntaxFacts.IsUsingAliasDirective(node))
+            {
+                syntaxFacts.GetPartsOfUsingAliasDirective(node, out _, out var aliasToken, out var name);
+                if (!aliasToken.Span.IsEmpty)
+                {
+                    // Use 'name.Parent' because VB can't resolve the declared symbol directly from 'node'
+                    var symbol = semanticModel.GetDeclaredSymbol(name.GetRequiredParent(), cancellationToken);
+                    if (IsSymbolObsolete(symbol))
+                        result.Add(aliasToken.Span);
+                }
+
+                return aliasToken;
+            }
+            else if (syntaxFacts.IsImplicitObjectCreationExpression(node))
+            {
+                syntaxFacts.GetPartsOfImplicitObjectCreationExpression(node, out var creationKeyword, out _, out _);
+                if (!creationKeyword.Span.IsEmpty)
+                {
+                    var symbol = semanticModel.GetSymbolInfo(node, cancellationToken).Symbol;
+                    if (IsSymbolObsolete(symbol) || IsSymbolObsolete(symbol?.ContainingType))
+                        result.Add(creationKeyword.Span);
+                }
+            }
+
+            return default;
+        }
+
+        void ProcessToken(SemanticModel semanticModel, SyntaxToken token)
+        {
+            if (syntaxFacts.IsIdentifier(token))
+            {
+                ProcessIdentifier(semanticModel, token);
+            }
+            else if (token.RawKind == _dimKeywordKind)
+            {
+                ProcessDimKeyword(result, semanticModel, token, cancellationToken);
+            }
+        }
+
+        void ProcessIdentifier(SemanticModel semanticModel, SyntaxToken token)
+        {
+            if (syntaxFacts.IsDeclaration(token.Parent))
+            {
+                var symbol = semanticModel.GetDeclaredSymbol(token.Parent, cancellationToken);
+                if (IsSymbolObsolete(symbol))
+                    result.Add(token.Span);
+            }
+            else
+            {
+                var symbol = semanticModel.GetSymbolInfo(token, cancellationToken).Symbol;
+                if (IsSymbolObsolete(symbol))
+                    result.Add(token.Span);
+            }
+        }
+    }
+
+    protected static bool IsSymbolObsolete([NotNullWhen(true)] ISymbol? symbol)
+    {
+        // Avoid infinite recursion. Iteration limit chosen arbitrarily; cases are generally expected to complete on
+        // the first iteration or fail completely.
+        for (var i = 0; i < 5; i++)
+        {
+            if (symbol is IAliasSymbol alias)
+            {
+                symbol = alias.Target;
+                continue;
+            }
+
+            if (symbol is INamedTypeSymbol { OriginalDefinition.SpecialType: SpecialType.System_Nullable_T, TypeArguments: [var valueType] })
```

Comment:
nit.  go whole hog and have a SymbolVisitor for this.  I think we should handle things like arrays or tuples for this as well.  

Basically, if the type signature (Which is not hard to traverse) has any types that are obsolete, it's fine to mark.  I would not just limt this to nullable.

Reply:
I thought about it, but decided against it in the end. I'll make sure there are tests added that indicate that they are simply reflecting current behavior and not necessarily one right or wrong answer.

Reply:
⏱️ Still need to add tests for arrays and tuples